mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala

```diff
@@ -100,6 +100,50 @@ sealed trait Matrix extends Serializable {
    *          corresponding value in the matrix with type `Double`.
    */
   private[spark] def foreachActive(f: (Int, Int, Double) => Unit)
+
+  override def hashCode(): Int = {
+    var result: Int = numRows * numCols + 31
+    this.foreachActive { case (rowInd, colInd, value) =>
+      // ignore explict 0 for comparison between sparse and dense
+      if (value != 0) {
+        result = 31 * result + rowInd + (numRows * colInd)
+        // refer to {@link java.util.Arrays.equals} for hash algorithm
+        val bits = java.lang.Double.doubleToLongBits(value)
+        result = 31 * result + (bits ^ (bits >>> 32)).toInt
+      }
+    }
+    result
```

Comment:
Could you please verify this?
There might be cases where a CSR matrix is the same as a CSC matrix. See (https://github.com/apache/spark/pull/5081/files#diff-8fbb9a5e1adf997a37f4d05521b8a1acR488)
If I hash `isTransposed` I would give different hashes for both of these, which are not desirable.